tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-InProc.cpp

```diff
@@ -0,0 +1,883 @@
+//===--- sourcekitdAPI-InProc.cpp -----------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "DictionaryKeys.h"
+#include "sourcekitd/sourcekitd.h"
+#include "sourcekitd/Internal.h"
+#include "sourcekitd/RequestResponsePrinterBase.h"
+#include "sourcekitd/CodeCompletionResultsArray.h"
+#include "sourcekitd/DocSupportAnnotationArray.h"
+#include "sourcekitd/TokenAnnotationsArray.h"
+#include "sourcekitd/Logging.h"
+#include "SourceKit/Core/LLVM.h"
+#include "SourceKit/Support/UIdent.h"
+#include "swift/Basic/ThreadSafeRefCounted.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/IntrusiveRefCntPtr.h"
+
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/MemoryBuffer.h"
+
+#include <map>
+
+using namespace SourceKit;
+using namespace sourcekitd;
+
+namespace {
+class SKDObject;
+typedef RefPtr<SKDObject> SKDObjectRef;
+
+class SKDObject : public ThreadSafeRefCountedBaseVPTR {
+public:
+  enum class ObjectKind {
+    Dictionary,
+    Array,
+    String,
+    Int64,
+    UID,
+    Bool,
+    CustomData,
+    Error,
+  };
+private:
+  ObjectKind Kind;
+protected:
+  SKDObject(ObjectKind K) : Kind(K) {}
+public:
+  SKDObject(SKDObject const&) = delete;
+  SKDObject &operator=(SKDObject const&) = delete;
+
+  ObjectKind getKind() const { return Kind; }
+
+  virtual sourcekitd_variant_type_t getVariantType() const = 0;
+
+  virtual void set(sourcekitd_uid_t Key, SKDObjectRef Value) {}
+  virtual void set(size_t Index, SKDObjectRef Value) {}
+
+  virtual SKDObjectRef get(sourcekitd_uid_t Key) const { return nullptr; }
+
+  virtual SKDObjectRef get(size_t Index) const { return nullptr; }
+  virtual size_t getCount() const { return 0; }
+
+  virtual sourcekitd_uid_t getUID() const { return nullptr; }
+  virtual Optional<int64_t> getInt64() const { return Optional<int64_t>(); }
```

Comment:
Ahhh nice. I was looking for something like that but couldn't track it down before.